crates/containerd-shim-wasm/src/sandbox/containerd.rs

```diff
@@ -0,0 +1,196 @@
+        let resp: Result<Vec<u8>> = self.rt.block_on(async {
+            let req = ReadContentRequest {
+                digest,
+                offset: 0,
+                size: 0,
+            };
+            let req = with_namespace!(req, self.namespace);
+            let response: tonic::Response<Streaming<ReadContentResponse>> = self
+                .content()
+                .read(req)
+                .await
+                .map_err(|err| ShimError::Others(err.to_string()))?;
+            let mut resp_stream = response.into_inner();
+
+            let mut data = vec![];
+            while let Some(mut next_message) = resp_stream
+                .message()
+                .await
+                .map_err(|err| ShimError::Others(err.to_string()))?
+            {
+                data.append(&mut next_message.data);
+            }
+
+            Ok(data)
+        });
+
+        resp
+    }
+
+    pub fn get_image_content_sha(&mut self, image_name: String) -> Option<String> {
+        let resp: Result<Option<String>> = self.rt.block_on(async {
+            let req = GetImageRequest { name: image_name };
+            let req = with_namespace!(req, self.namespace);
+            let resp = self
+                .images()
+                .get(req)
+                .await
+                .map_err(|err| ShimError::Others(err.to_string()))?;
+
+            let image_response = resp.into_inner();
+            match image_response.image {
+                Some(i) => {
+                    let desc = i.target.unwrap();
+                    Ok(Some(desc.digest))
+                }
+                None => Ok(None),
+            }
+        });
+        match resp {
+            Ok(layer_option) => layer_option,
+            Err(_) => None,
+        }
+    }
+
+    pub fn get_oci_image(&mut self, container_name: String) -> Option<String> {
+        let resp: Result<Option<String>> = self.rt.block_on(async {
+            let req = GetContainerRequest { id: container_name };
+            let req = with_namespace!(req, self.namespace);
+            let resp = self
+                .containers()
+                .get(req)
+                .await
+                .map_err(|err| ShimError::Others(err.to_string()))?;
+
+            let container_response = resp.into_inner();
+            match container_response.container {
+                Some(c) => Ok(Some(c.image)),
+                None => Ok(None),
+            }
+        });
+
+        match resp {
+            Ok(layer_option) => layer_option,
+            Err(_) => None,
+        }
+    }
+
+    // load module will query the containerd store to find an image that has an ArtifactType of WASM OCI Artifact
+    // If found it continues to parse the manifest and return the layers that contains the WASM modules
+    // and possibly other configuration artifacts
+    pub fn load_modules(&mut self, containerd_id: String) -> Option<Vec<oci::OciArtifact>> {
```

Comment:
Many of the fn's in this file return option and log warnings. I'm wondering if it would be better to return a result and bubble the error up rather than logging. WDYT?

Reply:
I considered this.  I was trying to consider the failure state, we don't know if this is artifact until we've done some basic parsing.  

If it isn't an artifact or we fail to gather the info from contianerd, should we hard error and not start the container or should we continue on and try to load like it is a container as a best effort (not an artifact)?

I opted to continue on as if it wasn't a container image since current validation checks would fail and the container would restart. So if there was a temporary reason, we weren't able to gather the correct information it would get tried again.

I can see the other approach here, is if we can't gather the info then we should fail fast.  thoughts?

Reply:
Thank you for explaining. I get now that the executor is optionally being populated with the artifact so that in a later stage it can decide how to load with best effort.

I don't know that you have to fail fast. I'm suggesting returning a typed error to the caller so the caller would have the information to decide to fail, log, or do nothing. It's more a nit about hiding errors.

I was suggesting a signature of:
```
pub fn load_modules(&mut self, containerd_id: String) -> Result<Option<Vec<oci::OciArtifact>>>
```

Then instance.rs would be responsible for logging or handling that error. You could still pass in the None to the executor. However, it seems a bit odd to pass in None to the executor, but I guess you have to get the artifact into the executor somehow.

Reply:
> I don't know that you have to fail fast. I'm suggesting returning a typed error to the caller so the caller would have the information to decide to fail, log, or do nothing. It's more a nit about hiding errors.

Make sense, I'll make this update